File: js/figures/figure7_1.js
"use strict";

// Port of code_distribute/python/Figure7_1.py
window.figureLib = window.figureLib || {};
window.figureLib.figure7_1 = function (outputGrid, params) {
  const nSampleA = Math.max(1, Math.round(params.n_sample_a));
  const nSampleB = Math.max(1, Math.round(params.n_sample_b));
  const sBar = Math.max(1, Math.round(params.s_bar));

  const nX = 100; // n_x - number of x-grid for plot
  const xP = Array.from({ length: nX }, (_, i) => i / (nX - 1)); // x grid points for plot
  const nF = 10; // number of features

  // Feature mapping phi(x) = [1, x, x^2, ..., x^9]
  function phi(x) {
    const out = new Array(nF);
    let p = 1;
    for (let k = 0; k < nF; k++) {
      out[k] = p;
      p *= x;
    }
    return out;
  }

  // dot(paramVec, phi(x)) evaluated over an array of x values
  function evalCurve(paramVec, xs) {
    return xs.map((x) => {
      const ph = phi(x);
      let s = 0;
      for (let k = 0; k < nF; k++) s += paramVec[k] * ph[k];
      return s;
    });
  }

  function fTrue(x) {
    return 2 * Math.sin(5 * x);
  }

  // Pad the observed min/max of a set of curves (and possibly extra points)
  // by ~10% to get a sensible ylim, since the original script leaves ylim
  // to matplotlib's auto-ranging.
  function computeYlim(curves, extraPoints, padFrac) {
    let lo = Infinity,
      hi = -Infinity;
    for (const c of curves) {
      for (const v of c) {
        if (v < lo) lo = v;
        if (v > hi) hi = v;
      }
    }
    if (extraPoints) {
      for (const v of extraPoints) {
        if (v < lo) lo = v;
        if (v > hi) hi = v;
      }
    }
    if (!isFinite(lo) || !isFinite(hi)) {
      lo = -1;
      hi = 1;
    }
    const pad = (hi - lo) * (padFrac !== undefined ? padFrac : 0.1) || 1;
    return [lo - pad, hi + pad];
  }

  // ---------------------------------------------------------------
  // Figure 7.1(a): prior sample functions f(x) = para . phi(x),
  // para ~ N(0, I_10)
  // ---------------------------------------------------------------
  {
    const zeroCurve = new Array(nX).fill(0);
    const sampleCurves = [];
    for (let i = 0; i < nSampleA; i++) {
      const para = window.rnd.randnVec(nF);
      sampleCurves.push(evalCurve(para, xP));
    }
    const ylim = computeYlim(sampleCurves.concat([zeroCurve]));

    const body = window.plotlib.makeCard(outputGrid, "Figure 7.1(a) — Prior sample functions");
    const chart = window.plotlib.createChart(body, {
      xlim: [0, 1],
      ylim: ylim,
      xlabel: "\\rm x",
      ylabel: "f({\\rm x})",
    });

    chart.line(xP, zeroCurve, { lineWidth: 3 });
    for (const fx of sampleCurves) {
      chart.line(xP, fx, { lineWidth: 0.5, color: [0.7, 0.7, 0.7] });
    }
    chart.finish();
  }

  // ---------------------------------------------------------------
  // Figure 7.1(b): Bayesian linear regression posterior, three noise
  // variance settings sigma_sq = [0.5, 100, 1e-6] (fixed, not parametrized)
  // ---------------------------------------------------------------
  {
    const sigmaSq = [0.5, 100, 1e-6];

    // x_data: s_bar evenly spaced points over [0,1] (matches Python's
    // `x = np.linspace(0, 1, s_bar)` used as the data locations).
    const xData = Array.from({ length: sBar }, (_, i) => (sBar === 1 ? 0 : i / (sBar - 1)));

    // Phi: n_f x s_bar (features as ROWS, data points as COLUMNS) --
    // matches numpy's np.column_stack([phi(x_s) for x_s in x]) shape exactly.
    const Phi = Array.from({ length: nF }, (_, k) => xData.map((x) => phi(x)[k]));
    const PhiT = window.linalg.transpose(Phi); // s_bar x n_f

    const y = xData.map((x) => fTrue(x) + window.rnd.randn());

    const means = [];
    const covs = [];
    for (let l = 0; l < sigmaSq.length; l++) {
      const A = window.linalg.matMul(PhiT, Phi); // s_bar x s_bar  (Phi.T @ Phi)
      const Areg = window.linalg.addDiag(A, sigmaSq[l]); // + sigma_sq[l] * I
      const Ainv = window.linalg.inv(Areg);
      const PhiAinv = window.linalg.matMul(Phi, Ainv); // n_f x s_bar
      const tmp0 = window.linalg.matMul(PhiAinv, PhiT); // n_f x n_f = Phi @ Ainv @ Phi.T
      const eyeN = window.linalg.eye(nF);
      const tmp = eyeN.map((row, i) => row.map((v, j) => v - tmp0[i][j]));
      const tmpT = window.linalg.transpose(tmp);
      const cov = tmp.map((row, i) => row.map((v, j) => (v + tmpT[i][j]) / 2));
      const mean = window.linalg.matVec(Phi, window.linalg.matVec(Ainv, y)); // Phi @ Ainv @ y

      means.push(mean);
      covs.push(cov);
    }

    const meanCurves = [
      evalCurve(means[0], xP),
      evalCurve(means[1], xP),
      evalCurve(means[2], xP),
    ];

    const sampleCurves = [];
    for (let i = 0; i < nSampleB; i++) {
      const para = window.rnd.mvnSample(means[0], covs[0]);
      sampleCurves.push(evalCurve(para, xP));
    }

    const ylim = computeYlim(meanCurves.concat(sampleCurves), y);

    const body = window.plotlib.makeCard(outputGrid, "Figure 7.1(b) — Posterior sample functions");
    const chart = window.plotlib.createChart(body, {
      xlim: [0, 1],
      ylim: ylim,
      xlabel: "\\rm x",
      ylabel: "f({\\rm x})",
      legendLoc: "upper right",
    });

    chart.line(xP, meanCurves[0], { lineWidth: 2, color: "blue", label: "\\sigma=0.5" });
    chart.line(xP, meanCurves[1], { lineWidth: 2, color: "orange", label: "\\sigma=10" });
    chart.line(xP, meanCurves[2], { lineWidth: 2, color: "green", label: "\\sigma=10^{-3}" });

    for (const fx of sampleCurves) {
      chart.line(xP, fx, { lineWidth: 0.3, color: [0.7, 0.7, 0.7] });
    }

    chart.scatter(xData, y, { marker: "circle", color: "blue", filled: false, label: "{\\rm y}_s" });

    chart.finish();
  }
};

File: python/Figure7_1.py
# Author: Kenji Kashima
# Date  : 2025/09/11

import numpy as np
import matplotlib.pyplot as plt
import config

np.random.seed(24)

n_x = 100;          # n_x - number of x-grid for plot
x_p = np.linspace(0,1,n_x)  # x grid points for plot

def phi(x:float)->np.ndarray:
    '''
        Feature mapping
        x - input
    '''
    return  np.array([1, x, x**2, x**3, x**4, 
                      x**5, x**6, x**7, x**8, x**9 ])

n_f = phi(0).size   # number of features
Phi_p = np.column_stack([phi(xi) for xi in x_p])    # matrix phi for plot

def f_true(x:float)->float:
    '''
        x - input
    '''
    return 2*np.sin(5*x)

def figure7_1a(n_sample:int=20):
    '''
        n_sample - number of sample functions
    '''
    figsize = config.global_config(type=1)
    plt.figure(figsize=figsize)
    plt.plot(x_p, np.zeros_like(x_p), linewidth=3)

    for _ in range (n_sample):
        para = np.random.randn(n_f)      
        fx =  para @ Phi_p                 
        plt.plot(x_p,fx,linewidth= 0.5, color=[0.7,0.7,0.7])    
    
    plt.xlabel(r'$\rm x$')
    plt.ylabel(r'$f({\rm x})$')
    plt.xlim([0,1])
    plt.grid()
    plt.tight_layout()
    plt.savefig("./Figure7_1a.pdf")
    plt.show()


def figure7_1b(sigma_sq, n_sample:int=20, s_bar:int=8):
    '''
        sigma_sq - list of squared SD
        n_sample - number of sample functions
        s_bar - number of data
    '''
    
    x = np.linspace(0,1,s_bar)     # x for data
    Phi = np.column_stack([phi(x_s) for x_s in x])    # matrix phi in (7.10)
    y = np.array([f_true(x_s) + np.random.randn() for x_s in x])  # noisy observation

    size_sigma_sq = len(sigma_sq)
    mean = np.zeros([n_f,size_sigma_sq])
    cov = np.zeros([n_f,size_sigma_sq*n_f])

    for l in range(size_sigma_sq):
        tmp = np.eye(n_f) - Phi @ np.linalg.inv(Phi.T @ Phi + sigma_sq[l] * np.eye(s_bar)) @ Phi.T
        cov[:,n_f*l:n_f*(l+1)] = (tmp+tmp.T)/2
        mean[:,l]  = Phi @ np.linalg.inv(Phi.T @ Phi + sigma_sq[l] * np.eye(s_bar)) @ y.T

    figsize = config.global_config(type=1)
    plt.figure(figsize=figsize)

    plt.plot(x_p, (mean[:,0] @ Phi_p).flatten(), linewidth= 2,label=r'$\sigma=0.5$',zorder=10)
    plt.plot(x_p, (mean[:,1] @ Phi_p).flatten(), linewidth= 2,label=r'$\sigma=10$')
    plt.plot(x_p, (mean[:,2] @ Phi_p).flatten(), linewidth= 2,label=r'$\sigma=10^{-3}$')

    # Sampling from the posterior
    for _ in range (n_sample):
        para = np.random.multivariate_normal(mean[:,0],cov[:,0:n_f])      
        fx =  para @ Phi_p                   
        plt.plot(x_p,fx,linewidth= 0.3,color=[0.7,0.7,0.7])

    plt.scatter(x,y,marker='o',label=r'${\rm y}_s$')
    plt.legend()
    
    plt.xlabel(r'$\rm x$')
    plt.ylabel(r'$f({\rm x})$')
    plt.xlim([0,1])
    plt.grid()
    plt.tight_layout()
    plt.savefig("./Figure7_1b.pdf")
    plt.show()

if __name__ == '__main__':
    figure7_1a(n_sample=20)

    sigma_sq = [0.25, 100, 10**(-6) ]  # standard deviation sigma = 0.5, 10, 10^{-3}

    figure7_1b(sigma_sq, n_sample=20, s_bar=20)
